cpp/src/sampling/random_walks.cuh

```diff
@@ -321,6 +286,28 @@ struct col_indx_extract_t<graph_t, index_t, std::enable_if_t<graph_t::is_multi_g
       [] __device__(auto crt_out_deg) { return crt_out_deg > 0; });
   }
 
+  // singleton for handling the case of missing
+  // weights, in which case a device buffer filled with
+  // `weight_t{1}` should be used as scratch-pad;
+  // this is static data to avoid device memory fragmentation
+  // by (potentially needlesly) re-allocating this scratch-pad;
+  //
+  static weight_t const* get_weights_scratch_pad(index_t num_edges, raft::handle_t const& handle)
```

Comment:
I agree. Unfortunately there's no common base type between `thrust::constant_iterator<T>` and `T*` (that I know of). So, what type should `values_` be? (currently it's `weight_t const*`). I have to think about other ways to avoid that allocation.

Reply:
I wonder, now that we're in C++ 17, if you can use `std::optional` to address this?  Pass the weight as an `std::optional<T const *>` and check in the code whether to use it or not?

Reply:
I deref `values_` in device. Although indirectly, via a capture `[...,values = values_]`. Maybe a `std::variant<weight_t const*, thrust::constant_iterator<weight_t>>` might work, but not in device...

Reply:
Certainly that allocation must go.

Reply:
How important is it to avoid testing for `values == nullptr` in the device code?  Not sure of your motivation for that concern.

You could make a real functor instead of a device lambda and use SFINAE to let the compiler optimize that away if you're concerned about that.

Reply:
No, I cannot use SFINAE on runtime values. `values == nullptr` is something decided at runtime, not compile-time.

Reply:
Yes, but you could dispatch to the right functor (use an if to pick the right implementation).

I know cudf has a number of examples of a templated device functor with a true/false in the template and anything within the `if (template_variable)` block is not compiled version where it's false and is in the version where it's true.  So you use the same code, compile it twice, and pick the right version at runtime.

Reply:
I'm trying to avoid the `if`. Not at any cost, though. If the workaround turns out to be overkill or hard to read, I'll just go with the `if` branching in device.

Reply:
The `if` outside the device function (to pick the right functor) is much lower cost.

Reply:
Is it? It's done in CPU. While inside the device lambda kernel it's done in GPU (many times...but in parallel).

Reply:
I believe it is if the graph is large enough.  It will get executed within each GPU thread multiple times, it would only get executed on the CPU once.  I suppose we could time it if we think the behavior is different.  But on a graph with a billion edges it feels like the if inside the device code would get executed thousands of times per thread which should be slower than executing it once on the CPU.

Anyway... however you think works best, I think either solution is better than allocating all of the extra memory.

Reply:
Done.